File: src/lib/scrapers/alternate.ts
import * as cheerio from "cheerio";
import type { PriceResult } from "../types";

const BASE = "https://www.alternate.nl";

export async function searchAlternate(query: string): Promise<PriceResult[]> {
  const url = `${BASE}/listing.xhtml?q=${encodeURIComponent(query)}&s=price_asc`;

  const res = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
      "Accept-Language": "nl-NL,nl;q=0.9",
    },
    signal: AbortSignal.timeout(10000),
    next: { revalidate: 300 },
  });

  if (!res.ok) throw new Error(`Alternate HTTP ${res.status}`);

  const html = await res.text();
  const $ = cheerio.load(html);
  const results: PriceResult[] = [];

  // Elke product-card is zelf een <a class="productBox">
  $("a.productBox").each((_, el) => {
    const name = $(el).find(".product-name").first().text().replace(/\s+/g, " ").trim();

    const href = $(el).attr("href") ?? "";
    const link = href.startsWith("http") ? href : `${BASE}${href}`;

    // span.price is de actuele prijs; doorgestreepte prijzen hebben geen .price-class
    const priceText = $(el)
      .find("span.price")
      .first()
      .text()
      .replace(/[^\d,]/g, "")
      .replace(",", ".");
    const price = parseFloat(priceText);

    const imgSrc = $(el).find("img.productPicture").first().attr("src");
    const img = imgSrc
      ? imgSrc.startsWith("http")
        ? imgSrc
        : `${BASE}${imgSrc}`
      : undefined;

    const delivery = $(el).find(".delivery-info").text().toLowerCase();
    const inStock = delivery.includes("op voorraad");

    if (name && !isNaN(price) && price > 0) {
      results.push({ retailer: "alternate", name, priceEur: price, url: link, imageUrl: img, inStock });
    }
  });

  return results.slice(0, 10);
}

File: scrapers/corebuild_scrapers/retailers/alternate.py
"""Alternate: listing.xhtml — elke product-card is zelf een <a class="productBox">."""

from urllib.parse import quote

from bs4 import BeautifulSoup

from .common import MAX_RESULTS, get, parse_nl_price

BASE = "https://www.alternate.nl"


def search(query: str) -> list[dict]:
    res = get(f"{BASE}/listing.xhtml?q={quote(query)}&s=price_asc")
    soup = BeautifulSoup(res.text, "html.parser")
    results = []

    for card in soup.select("a.productBox"):
        name_el = card.select_one(".product-name")
        price_el = card.select_one("span.price")
        img = card.select_one("img.productPicture")
        delivery = card.select_one(".delivery-info")

        if not name_el or not price_el:
            continue
        price = parse_nl_price(price_el.get_text())
        if price is None:
            continue

        href = card.get("href", "")
        img_src = img.get("src", "") if img else ""
        in_stock = "op voorraad" in (delivery.get_text(" ").lower() if delivery else "")

        results.append(
            {
                "retailer": "alternate",
                "name": " ".join(name_el.get_text(" ", strip=True).split()),
                "price_eur": price,
                "url": href if href.startswith("http") else f"{BASE}{href}",
                "image_url": (img_src if img_src.startswith("http") else f"{BASE}{img_src}") if img_src else None,
                "in_stock": in_stock,
            }
        )
        if len(results) >= MAX_RESULTS:
            break

    return results
